File: ROS/src/simple_ros_cpp/src/service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"
#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono_literals;

class ServiceClient : public rclcpp::Node
{
public:
    ServiceClient()
        : Node("service_client")
    {
        _client = create_client<std_srvs::srv::SetBool>("setBool");

        while (!_client->wait_for_service(1s))
        {
            RCLCPP_INFO(get_logger(), "service not available");
        }
        _request = std::make_shared<std_srvs::srv::SetBool::Request>();
        _send_timer = create_wall_timer(3s, std::bind(&ServiceClient::send_request, this));
        _update_timer = create_wall_timer(100ms, std::bind(&ServiceClient::update, this));
        // send_request();
    }

private:
    rclcpp::Client<std_srvs::srv::SetBool>::SharedPtr _client;
    std::shared_ptr<std_srvs::srv::SetBool::Request> _request;
    rclcpp::TimerBase::SharedPtr _send_timer;
    rclcpp::TimerBase::SharedPtr _update_timer;
    bool _bool;
    void send_request()
    {
        // _request <- data
        if (!_request->data)
        {
            _request->data = true;
        }
        else
        {
            _request->data = false;
        }

        auto future = _client->async_send_request(_request,
                                                  std::bind(&ServiceClient::done_callback,
                                                            this,
                                                            std::placeholders::_1));
    }
    void done_callback(rclcpp::Client<std_srvs::srv::SetBool>::SharedFuture future)
    {
        auto response = future.get();
        RCLCPP_INFO(get_logger(), response->message.c_str());
        RCLCPP_INFO(get_logger(), "%s", response->success ? "true" : "false");
    }
    void update()
    {
        RCLCPP_INFO(get_logger(), "updating now !!!");
    }
};

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<ServiceClient>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ROS/src/simple_ros/simple_ros/service_client.py
import rclpy
from rclpy.node import Node
from std_srvs.srv import SetBool


class Service_client(Node):
    def __init__(self):
        super().__init__("service_client")
        self.client = self.create_client(SetBool, "setBool")
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("service not available")
        self.request = SetBool.Request()
        self.create_timer(0.1, self.update)
        # self.create_timer(3, self.send_reqarithmetic_resultuest)
        self.send_request()

    def send_request(self):
        self.future = self.client.call_async(self.request)
        self.future.add_done_callback(self.done_callback)

    def done_callback(self, future):
        response : SetBool.Response = future.result()
        self.get_logger().info(f"{response.success}")
        self.get_logger().info(response.message)

    def update(self):
        self.get_logger().info("updating!!")

def main():
    rclpy.init()
    node = Service_client()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.destroy_node()

if __name__ == "__main__":
    main()
